src/Auxiliary/Aux_Check_Parameter.cpp

```
@@ -1159,8 +1172,38 @@ void Aux_Check_Parameter()
    if ( OPT__INIT == INIT_BY_FILE )
       Aux_Message( stderr, "WARNING : currently we don't check MIN_DENS for the initial data loaded from UM_IC !!\n" );
 
-   } // if ( MPI_Rank == 0 )
 
+#  elif ( WAVE_SCHEME == WAVE_GRAMFE ) // #  if ( WAVE_SCHEME == WAVE_FD )
+
+   const double dt_fluid_max = 0.4;
+
+   if ( DT__FLUID > dt_fluid_max )
+      Aux_Message( stderr, "WARNING : %s solver with DT__FLUID (%13.7e) > %13.7e is unstable !!\n",
+                   "WAVE_GRAMFE", DT__FLUID, dt_fluid_max );
+
+#  if ( SUPPORT_FFTW == FFTW3 && defined(GRAMFE_ENABLE_SINGLE_PRECISION) )
+   Aux_Message( stderr, "WARNING : %s solver using GRAMFE_ENABLE_SINGLE_PRECISION is unstable !!\n",
+                  "WAVE_GRAMFE"  );
+#  endif // #  if ( SUPPORT_FFTW == FFTW3 && defined(GRAMFE_ENABLE_SINGLE_PRECISION) )
+
+#  if ( SUPPORT_FFTW == FFTW2 && !(defined(FLOAT8)) )
+   Aux_Message( stderr, "WARNING : %s solver using CPU and FFTW2 without FLOAT8 is unstable !!\n",
+                  "WAVE_GRAMFE"  );
+#  endif // #  if ( SUPPORT_FFTW == FFTW2 && !(defined(FLOAT8)) )
+
+#  if ( SUPPORT_FFTW == FFTW2 && defined(GRAMFE_ENABLE_SINGLE_PRECISION) )
+   Aux_Message( stderr, "WARNING : %s solver using FFTW2 ignores GRAMFE_ENABLE_SINGLE_PRECISION !!\n",
+                  "WAVE_GRAMFE"  );
+#  endif // #  if ( SUPPORT_FFTW == FFTW2 && defined(GRAMFE_ENABLE_SINGLE_PRECISION) )
```

IIUC, there are two different constraints here.
1. `GRAMFE_ENABLE_SINGLE_PRECISION` is always unstable no matter using CPU (`FFTW2, FFTW3`) or GPU (`GRAMFE_ENABLE_GPU`). If so, should we raise an error instead of a warning?
2. For `( SUPPORT_FFTW == FFTW2  &&  !defined GRAMFE_ENABLE_GPU )`, `FLOAT8` and `GRAMFE_ENABLE_SINGLE_PRECISION` are mutually exclusive. We should raise an error if this condition is violated.

If you agree with it, please revise the current three checks to better distinguish these two constraints.

---

I agree and will follow your advice on 2. If you think that we should just define `GRAMFE_FLOAT8` by default, we can just raise an error if it is not defined maybe?

---

Yes. Just raise an error when disabling `GRAMFE_FLOAT8`.